Adding cryptography-compatible pyOpenSSL to source dist

benjaoming opened the pull request:
<!--
 1. Following guidance below, replace …'s with your own words
 2. After saving the PR, tick of completed checklist items
 3. Skip checklist items that are not applicable or not necessary
 4. Delete instruction/comment blocks
-->

### Summary

Intended as a:

 - **final solution** for the sdist that carries bundled in compiled dependencies for the most popular platforms
 - **temporary solution** in the sense that it fixes an issue caused by having a platform-specific dependency that may interact with the rest of the system, since we cannot rule out that some other system library would expect cryptography x.y.z and find our version 2.0.3

#### Fixing `SNIMissingWarning` ?

To reproduce the error that this PR targets, do the following:

1. Run the Pex provided in our latest release on a system that already has pyOpenSSL
1. `kolibri manage shell`
1. @lyw07 would it be sufficient to just run `requests.get("https://studio.learningequality.org")` to reproduce?


#### Fixing `requests` depends on pyOpenSSL (bindings to system's OpenSSL) which expects system's version of cryptography but gets ours

I don't even rembember what the error was, but it was something with `_lib.SSL_ST_INIT` and some `AttributeError` - @lyw07 ?

To reproduce the error that this PR targets, do the following:

1. Run the Pex provided in our latest release on a system that already has pyOpenSSL
1. `kolibri manage shell`
1. @lyw07 need help here :)

### Reviewer guidance

Is this latest version of pyOpenSSL adequate? Will it work on all platforms?

### References

#2544 

----

### Contributor Checklist

- [x] Test on Ubuntu 14.04
- [x] Test on Ubuntu 16.04
- [x] Test on Ubuntu 17.10
- [x] PR has the correct target branch and milestone
- [x] PR has 'needs review' or 'work-in-progress' label
- [x] Contributor has fully tested the PR manually
- [x] Screenshots of any front-end changes are in the PR description
- [x] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')

### Reviewer Checklist

- [ ] Automated test coverage is satisfactory
- [x] Reviewer has fully tested the PR manually
- [ ] PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- [ ] External dependencies files were updated (`yarn` and `pip`)
- [ ] Documentation is updated
- [ ] Link to diff of internal dependency change is included
- [ ] CHANGELOG.rst is updated for high-level changes
- [ ] Contributor is in AUTHORS.rst

benjaoming commented on Makefile:
```diff
@@ -56,7 +57,7 @@ test:
 	tox
 
-assets: staticdeps
+assets:
```
I removed `staticdeps` from `make assets` because I wanted to be able to quickly build just the "assets" as in JS, CSS etc without getting `kolibri/dist` populated... it's confusing once it's populated because when installing other versions with `pip install` or switching branches, the contents of this folder remain unchanged and override whatever is installed in the `virtualenv`.

lyw07 commented:
Instead of using the latest 0.7 release pex, I think we only get the error when we install the latest whl file in release-0.7.x branch, because:
1. 0.7 release doesn't have your PR that fixes the order of PYTHONPATH, so it will produce another error I believe
2. The pex file will not produce the error because `OpenSSL` is in `/usr/lib/python2.7/disk-packages/`, which is not in Kolibri's PYTHONPATH. 

Sorry now I feel like it might not be the problem of SNI, but just `pyopenssl` in the system is not compatible with our bundled `cryptography`, since importing a channel is working in pex file with python 2.7.6. 


To produce the error after installing the latest whl file from `release-0.7.x` branch, I think what you said is right! We just need:
1. run `kolibri shell`
2. import `request`
3. requests.get("https://studio.learningequality.org")

to get the following error:
```
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/usr/local/lib/python2.7/dist-packages/kolibri/dist/requests/api.py", line 72, in get
    return request('get', url, params=params, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/kolibri/dist/requests/api.py", line 58, in request
    return session.request(method=method, url=url, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/kolibri/dist/requests/sessions.py", line 508, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/local/lib/python2.7/dist-packages/kolibri/dist/requests/sessions.py", line 618, in send
    r = adapter.send(request, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/kolibri/dist/requests/adapters.py", line 506, in send
    raise SSLError(e, request=request)
SSLError: HTTPSConnectionPool(host='studio.learningequality.org', port=443): Max retries exceeded with url: / (Caused by SSLError(SSLError(1, '_ssl.c:510: error:14077410:SSL routines:SSL23_GET_SERVER_HELLO:sslv3 alert handshake failure'),))
```



And if we try to upgrade `pyopenssl` in Ubuntu 14.04 from 0.13 to versions >=0.14 but <16.2.0, or running a Ubuntu 16.04 VM, and then we do `kolibri shell`, we will get the following error:
```
Traceback (most recent call last):
  File "/usr/local/bin/kolibri", line 7, in <module>
    from kolibri.utils.cli import main
  File "/usr/local/lib/python2.7/dist-packages/kolibri/utils/cli.py", line 22, in <module>
    from . import server  # noqa
  File "/usr/local/lib/python2.7/dist-packages/kolibri/utils/server.py", line 8, in <module>
    import requests
  File "/usr/local/lib/python2.7/dist-packages/kolibri/dist/requests/__init__.py", line 84, in <module>
    from urllib3.contrib import pyopenssl
  File "/usr/local/lib/python2.7/dist-packages/kolibri/dist/urllib3/contrib/pyopenssl.py", line 46, in <module>
    import OpenSSL.SSL
  File "/usr/local/lib/python2.7/dist-packages/OpenSSL/__init__.py", line 8, in <module>
    from OpenSSL import rand, crypto, SSL
  File "/usr/local/lib/python2.7/dist-packages/OpenSSL/SSL.py", line 105, in <module>
    SSL_ST_INIT = _lib.SSL_ST_INIT
AttributeError: 'module' object has no attribute 'SSL_ST_INIT'
```

lyw07 commented on Makefile:
```diff
@@ -83,6 +84,7 @@ staticdeps:
 	git checkout -- kolibri/dist # restore __init__.py
 	pip install -t kolibri/dist -r $(REQUIREMENTS)
 	python install_cexts.py --file $(REQUIREMENTS_CEXT) # pip install c extensions
+	pip install -t kolibri/dist -r $(REQUIREMENTS_CEXT_NOARCH)
```
Is it possible to add --no-deps install option here? Since `pyopenssl` has dependencies such as `cryptography`, I think just installing it without dependencies will be enough for us?

benjaoming replied:
Ah yes, true, will fix this ASAP

lyw07 commented on requirements/cext_noarch.txt:
```diff
@@ -0,0 +1,4 @@
+# These are unique requirements for C extensions, but the
+# ones that don't need to be shipped in one version per
+# arch/OS/pyVesion
+pyOpenSSL==17.5.0
```
I think 17.5.0 will not work for us currently, since it requires `cryptography` to be >= 2.1.4. But we are using 2.0.3 right now. I tried yesterday to upgrade `pyopenssl` to 17.5.0, and there were some errors saying that "attribute xxx not found". So I think for now, it will be better to use `pyopenssl` 17.4.0 to be compatible with the current version of `cryptography` we are using?

benjaoming replied:
Thanks! Will do that right meow!

Yes, the upgrade path for `cryptography` starts by adding the 2.1.4 version to Morango's test matrix, and then we can upgrade it in Kolibri... so this has to wait.

benjaoming commented:
Thanks for clarifying @lyw07 ! re: Pex - I think it actually does have system paths, for instance `'/usr/lib/python2.7'` is present.

Alternatively, one can create a virtualenv with `--system-site-packages` or building and installing the `.deb`.. latter is a bit difficult just for troubleshooting this, so I would suggest the virtualenv way.

> 0.7 release doesn't have your PR that fixes the order of PYTHONPATH, so it will produce another error I believe

I guess it's the errors on the current `develop` branch (with the PYTHONPATH fixes) that are relevant. Those should still detect the system's OpenSSL, because it's just a reorder of the path, where the system paths have lower priority.

lyw07 commented:
@benjaoming Sorry I'm still confused about the pex file. After downloading the latest pex file from release-0.7.x branch and run `python kolibri.pex shell` to import `OpenSSL`, it says it's not able to find the module `OpenSSL`.. Could you explain a bit more of this? Thank you!!

lyw07 commented:
I'm testing the new whl file from this PR, but got this error when running `kolibri shell`:

```
Traceback (most recent call last):
  File "/usr/local/bin/kolibri", line 7, in <module>
    from kolibri.utils.cli import main
  File "/usr/local/lib/python2.7/dist-packages/kolibri/utils/cli.py", line 42, in <module>
    from . import server  # noqa
  File "/usr/local/lib/python2.7/dist-packages/kolibri/utils/server.py", line 6, in <module>
    import cherrypy
  File "/usr/local/lib/python2.7/dist-packages/kolibri/dist/cherrypy/__init__.py", line 66, in <module>
    from ._cperror import (
  File "/usr/local/lib/python2.7/dist-packages/kolibri/dist/cherrypy/_cperror.py", line 128, in <module>
    from jaraco.classes.properties import classproperty
ImportError: No module named jaraco.classes.properties
```
This error exists in the whl file and pex file of develop branch as well. Should I file an separate issue for this?

benjaoming commented:
@lyw07 that's weird, we did upgrade CherryPy, so it could be related. It's a new dependency that they introduced in 11.1.0 (we're on 13.0.0): https://github.com/cherrypy/cherrypy/commit/066c8e2b7ae890fee70ce3a3a0314c7e83574d90

indirectlylit commented:
Should we be working off the 0.7.x branch rather than develop, so that we can target 0.7.1 for Debian?

lyw07 commented:
@benjaoming It seems that if we install cherrypy to a directory by doing `pip install -t <directory> cherrypy`, `jaraco` will not be found. If we indeed install cherrypy directly to the system by doing `pip install cherrypy`, then we will be able to find `jaraco`.

benjaoming commented:
@lyw07 woah, I wasn't aware of that, I thought dependencies got resolved, we should add the `--no-deps` flag to our testing then.. if you do that, you can probably reproduce the issue and commit a fix by adding a missing dependency (or more?). Try using `pipdeptree` to introspect.

benjaoming commented:
@lyw07 I've rebased to current `develop` (with the CherryPy download)

lyw07 commented:
Tested on Ubuntu 16.04 and 17.10 as well, and the issue doesn't exist. 🎉
@benjaoming Is this PR still work-in-progress?

benjaoming commented:
@lyw07 

Almost done! I'm not sure how we track release notes, but am kind of worried that we don't communicate clearly enough how issues are resolved etc. while so many people are absent on holidays. Perhaps now is the time to introduce something :) I'll bring it up later, will merge once it seems clear.

benjaoming commented:
THANKS FOR TESTING!!

indirectlylit commented:
* Should we merge this?
* Should this be in 0.7.1?

benjaoming commented:
@indirectlylit I believe we should merge it.

Because of the long break, I am sort of loosing track of what was the accepted support for 0.7.x, stuff we already fixed in relation to path and C extension issues -- but I believe that we accepted that not all Ubuntu platforms (like 14.04) would be supported.

I also think we switched off SSL in 0.7.x, and I don't think we should backport this in order to re-introduce SSL, given the additional amount of testing and further unclarity. I think, if people are already running a 0.7 release, they would expect their platforms to continue to be supported, and we shouldn't worry too much about additional platform support since 0.8 isn't too far (especially not if we mean the first pre-releases).

benjaoming commented:
Re: My own words

> we accepted that not all Ubuntu platforms (like 14.04) would be supported

Just want to add, that my understanding is that 14.04 support will be re-introduced in 0.8.
